component/topsql/store/store.go

```diff
@@ -165,35 +172,97 @@ func topSQLProtoToMetric(
 }
 
 // transform resource_usage_agent.CPUTimeRecord to util.Metric
-func rsMeteringProtoToMetric(
+func rsMeteringProtoToMetrics(
 	instance, instance_type string,
 	record *rsmetering.ResourceUsageRecord,
-) (m Metric, err error) {
+) (ms []Metric, err error) {
 	tag := tipb.ResourceGroupTag{}
-
-	m.Metric.Name = "cpu_time"
-	m.Metric.Instance = instance
-	m.Metric.InstanceType = instance_type
-
 	tag.Reset()
 	if err = tag.Unmarshal(record.ResourceGroupTag); err != nil {
 		return
 	}
-
-	m.Metric.SQLDigest = hex.EncodeToString(tag.SqlDigest)
-	m.Metric.PlanDigest = hex.EncodeToString(tag.PlanDigest)
+	sqlDigest := hex.EncodeToString(tag.SqlDigest)
+	planDigest := hex.EncodeToString(tag.PlanDigest)
+
+	mCpu := Metric{
+		Metric: topSQLTags{
+			Name:         "cpu_time",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mReadRow := Metric{
+		Metric: topSQLTags{
+			Name:         "read_row",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mReadIndex := Metric{
+		Metric: topSQLTags{
+			Name:         "read_index",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mWriteRow := Metric{
+		Metric: topSQLTags{
+			Name:         "write_row",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mWriteIndex := Metric{
+		Metric: topSQLTags{
+			Name:         "write_index",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
 
 	for i := range record.RecordListCpuTimeMs {
 		tsMillis := record.RecordListTimestampSec[i] * 1000
 		cpuTime := record.RecordListCpuTimeMs[i]
 
-		m.Timestamps = append(m.Timestamps, tsMillis)
-		m.Values = append(m.Values, cpuTime)
+		mCpu.Timestamps = append(mCpu.Timestamps, tsMillis)
+		mCpu.Values = append(mCpu.Values, cpuTime)
+
+		appendMetricRowIndex(i, tsMillis, record.RecordListReadKeys, &mReadRow, &mReadIndex, tag.Label)
+		appendMetricRowIndex(i, tsMillis, record.RecordListWriteKeys, &mWriteRow, &mWriteIndex, tag.Label)
 	}
 
+	ms = append(ms, mCpu, mReadRow, mReadIndex, mWriteRow, mWriteIndex)
 	return
 }
 
+// appendMetricRowIndex only used in rsMeteringProtoToMetrics, just used to reduce repetition.
+func appendMetricRowIndex(i int, ts uint64, values []uint32, mRow, mIndex *Metric, label *tipb.ResourceGroupTagLabel) {
+	var rows, indexes uint32
+	if len(values) >= i {
```

Comment:
> I think they must be equal, because we use two lists to "simulate" the hashmap, which is guaranteed by TiKV: L124-L33

so `len(values) > i` always be true?

Reply:
Yes, this is just to prevent out of bound panic.

Reply:
I think it should be `len(values) > i` not `len(values) >= i`

Reply:
Yes! fixed